testng-core/src/test/java/test/retryAnalyzer/issue2798/HashCodeAwareRetryAnalyzer.java

```diff
@@ -4,17 +4,22 @@
 import java.util.List;
 import org.testng.IRetryAnalyzer;
 import org.testng.ITestResult;
+import org.testng.internal.AutoCloseableLock;
 
 public class HashCodeAwareRetryAnalyzer implements IRetryAnalyzer {
 
   public static final List<Integer> hashCodes = new ArrayList<>();
 
+  private static final AutoCloseableLock lock = new AutoCloseableLock();
```

Comment:
```suggestion
  private final AutoCloseableLock lock = new AutoCloseableLock();
```

Previously this lock was instance-based, so `static` should not be here

Reply:
I think the previous implementation was wrong. There cannot be a situation wherein two threads use the same retry analyser to invoke. So the retry method should never have been `synchronized`. The intent was to ensure that the static data member `hashCodes` which is being shared by multiple instances of the `HashCodeAwareRetryAnalyzer` all try and update the same `List` and so it should have been guarded with a lock.